File: weltwerk/verify/claim_ledger.py
# SPDX-License-Identifier: AGPL-3.0-only
"""
claim_ledger.py — a domain-agnostic, honesty-enforced claim ledger (extracted from the snowflake study).

A reusable template for stating a set of claims about ANY modeled system such that each claim is forced to
carry its epistemic status. Every claim has a GRADE on the ladder, the MECHANISM it rests on, what it does NOT
show, and a FALSIFIER; and every claim projects into the shared `AnalysisResult` honesty contract (scope + ≥1
limitation). `audit_ledger` mechanizes the ledger-level invariants — so a ledger cannot silently ship an
ungraded, unfalsifiable, or boundary-free claim. `integrity ≠ truth`; `salience ≠ importance`.

GRADE LADDER (preserve epistemic states; never force a binary):
  ESTABLISHED     — settled theory/observation, broadly reproduced.
  MEASURED        — a measured effect exists; full mechanism may be open.
  UNDERDETERMINED — competing explanations survive the evidence.
  SPECULATIVE     — proposed, not supported by current evidence.
  NOT_MEASURED    — outside what has been observed (a scope note).
"""
from __future__ import annotations

import os
import sys
from dataclasses import dataclass
from typing import Tuple

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from artifacts import AnalysisResult, Finding, Limitation        # noqa: E402  (honesty contract, reused)

GRADES = ("ESTABLISHED", "MEASURED", "UNDERDETERMINED", "SPECULATIVE", "NOT_MEASURED")
SUPPORTED = {"ESTABLISHED", "MEASURED"}      # grades that assert a claim is currently supported


@dataclass(frozen=True)
class Claim:
    id: str
    statement: str
    grade: str
    mechanism: str                # what the claim actually rests on
    does_not_show: str            # the boundary — what it does NOT establish
    falsifier: str                # the observation that would overturn it
    quantity: str = ""            # optional grounded number

    def as_analysis(self) -> AnalysisResult:
        """Project into the shared honesty contract (scope + ≥1 limitation)."""
        findings = (
            Finding("CLAIM", "claim-ledger", f"[{self.grade}] {self.statement}"),
            Finding("MECHANISM", "claim-ledger", self.mechanism),
        )
        limitations = (
            Limitation("claim-ledger", f"does not show: {self.does_not_show}"),
            Limitation("epistemic", f"grade={self.grade}; falsifier: {self.falsifier}"),
        )
        return AnalysisResult(source_trace=(self.id,), scope="claim-ledger",
                              findings=findings, limitations=limitations)


def grade_counts(ledger) -> dict:
    out = {g: 0 for g in GRADES}
    for c in ledger:
        out[c.grade] = out.get(c.grade, 0) + 1
    return out


def audit_ledger(ledger) -> dict:
    """Ledger-level honesty invariants. `honest` iff every grade is on the ladder and every claim carries both
    a non-empty falsifier and a non-empty 'does_not_show' boundary."""
    off_ladder = [c.id for c in ledger if c.grade not in GRADES]
    missing = [c.id for c in ledger if not (c.falsifier.strip() and c.does_not_show.strip())]
    return {"off_ladder": off_ladder, "missing_falsifier_or_boundary": missing,
            "honest": not off_ladder and not missing, "counts": grade_counts(ledger)}


def main():
    print("claim_ledger.py — a domain-agnostic, honesty-enforced claim ledger\n")
    demo = (
        Claim("D1", "Effect E is real and reproduced.", "ESTABLISHED",
              "mechanism M, reproduced across N studies.", "the magnitude under condition C.",
              "a pre-registered replication failing to find E."),
        Claim("D2", "Hypothesis H explains the residual.", "SPECULATIVE",
              "no established mechanism; consistent-with but not required-by the data.",
              "anything — listed to be graded, not endorsed.",
              "a controlled test isolating H from the known driver."),
    )
    a = audit_ledger(demo)
    print(f"  audit: honest={a['honest']}  counts={a['counts']}")
    for c in demo:
        print(f"    {c.id} [{c.grade}] falsifier? {'yes' if c.falsifier else 'NO'}")
    print("\n  every claim carries grade + mechanism + boundary + falsifier, and projects to AnalysisResult.")
    print("  a ledger cannot ship an ungraded, unfalsifiable, or boundary-free claim. integrity ≠ truth.")


if __name__ == "__main__":
    main()

File: Rust/src/claim_ledger.rs
// SPDX-License-Identifier: AGPL-3.0-only
//! The graded claim ledger. `Grade` is an enum, so an off-ladder grade is *unrepresentable* — a Rust
//! strengthening of the Python string ladder. A claim must carry the boundary it does NOT establish and the
//! observation that would falsify it; `audit_ledger` refuses a ledger missing either. `grade ≠ truth`;
//! `honest ≠ true`.

use std::collections::BTreeMap;

use crate::artifacts::{AnalysisResult, Finding, Limitation};
use crate::epistemic_types::Grounding;

/// The epistemic ladder. SUPPORTED = {Established, Measured}.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Grade {
    Established,
    Measured,
    Underdetermined,
    Speculative,
    NotMeasured,
}

impl Grade {
    pub const ALL: [Grade; 5] = [
        Grade::Established,
        Grade::Measured,
        Grade::Underdetermined,
        Grade::Speculative,
        Grade::NotMeasured,
    ];

    pub fn as_str(&self) -> &'static str {
        match self {
            Grade::Established => "ESTABLISHED",
            Grade::Measured => "MEASURED",
            Grade::Underdetermined => "UNDERDETERMINED",
            Grade::Speculative => "SPECULATIVE",
            Grade::NotMeasured => "NOT_MEASURED",
        }
    }

    /// Grades that assert a claim is currently supported.
    pub fn supported(&self) -> bool {
        matches!(self, Grade::Established | Grade::Measured)
    }
}

/// A graded, falsifiable claim. `does_not_show` and `falsifier` are REQUIRED to be non-empty for the ledger
/// to audit as honest.
#[derive(Debug, Clone)]
pub struct Claim {
    pub id: String,
    pub statement: String,
    pub grade: Grade,
    pub mechanism: String,
    pub does_not_show: String,
    pub falsifier: String,
    pub quantity: String,
}

impl Claim {
    pub fn new(
        id: impl Into<String>,
        statement: impl Into<String>,
        grade: Grade,
        mechanism: impl Into<String>,
        does_not_show: impl Into<String>,
        falsifier: impl Into<String>,
    ) -> Self {
        Self {
            id: id.into(),
            statement: statement.into(),
            grade,
            mechanism: mechanism.into(),
            does_not_show: does_not_show.into(),
            falsifier: falsifier.into(),
            quantity: String::new(),
        }
    }

    pub fn with_quantity(mut self, q: impl Into<String>) -> Self {
        self.quantity = q.into();
        self
    }

    /// Project into the shared honesty contract. Always honest by construction (scope + 2 limitations).
    pub fn as_analysis(&self) -> AnalysisResult {
        AnalysisResult::new(
            "claim",
            vec![
                Finding::new("CLAIM", "claim", self.statement.clone()),
                Finding::new("GRADE", "claim", self.grade.as_str()),
                Finding::new("MECHANISM", "claim", self.mechanism.clone()),
            ],
            vec![
                Limitation::new("claim", format!("does not show: {}", self.does_not_show)),
                Limitation::new("claim", format!("falsifier: {}", self.falsifier)),
            ],
        )
        .expect("claim as_analysis always satisfies the honesty contract")
    }
}

/// The result of auditing a ledger. `honest` requires every claim to carry both boundary fields. Grades are
/// on-ladder by construction (the `Grade` enum), so there is no off-ladder failure mode to check.
#[derive(Debug, Clone)]
pub struct LedgerAudit {
    pub honest: bool,
    pub counts: BTreeMap<&'static str, usize>,
    pub missing_boundary: Vec<String>,
}

pub fn grade_counts(claims: &[Claim]) -> BTreeMap<&'static str, usize> {
    let mut counts: BTreeMap<&'static str, usize> = BTreeMap::new();
    for g in Grade::ALL {
        counts.insert(g.as_str(), 0);
    }
    for c in claims {
        *counts.get_mut(c.grade.as_str()).unwrap() += 1;
    }
    counts
}

pub fn audit_ledger(claims: &[Claim]) -> LedgerAudit {
    let missing: Vec<String> = claims
        .iter()
        .filter(|c| c.does_not_show.trim().is_empty() || c.falsifier.trim().is_empty())
        .map(|c| c.id.clone())
        .collect();
    LedgerAudit { honest: missing.is_empty(), counts: grade_counts(claims), missing_boundary: missing }
}

/// Grounded iff the wrapped claim is graded ESTABLISHED/MEASURED on the ladder.
pub struct SupportedClaim<'a> {
    pub claim: &'a Claim,
}

impl<'a> Grounding for SupportedClaim<'a> {
    fn is_grounded(&self) -> bool {
        self.claim.grade.supported()
    }
    fn label(&self) -> String {
        format!("claim-grade={}", self.claim.grade.as_str())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn good(id: &str, g: Grade) -> Claim {
        Claim::new(id, "stmt", g, "mech", "the boundary", "a failing observation")
    }

    #[test]
    fn honest_ledger_passes() {
        let ledger = vec![good("A", Grade::Established), good("B", Grade::Measured)];
        assert!(audit_ledger(&ledger).honest);
    }

    #[test]
    fn missing_falsifier_is_dishonest() {
        let bad = Claim::new("X", "stmt", Grade::Measured, "mech", "boundary", "");
        let a = audit_ledger(&[bad]);
        assert!(!a.honest);
        assert_eq!(a.missing_boundary, vec!["X".to_string()]);
    }

    #[test]
    fn supported_claim_grounds_only_when_supported() {
        let est = good("A", Grade::Established);
        let spec = good("B", Grade::Speculative);
        assert!(SupportedClaim { claim: &est }.is_grounded());
        assert!(!SupportedClaim { claim: &spec }.is_grounded());
    }
}
